ecmascript/transforms/compat/src/es2015/arrow.rs

```diff
@@ -112,20 +148,85 @@ impl Fold for Arrow {
                     },
                 });
 
-                if !used_this {
+                if !used_this && !used_arguments {
                     return fn_expr;
                 }
 
+                let mut args = vec![ThisExpr { span: DUMMY_SP }.as_arg()];
+                if used_arguments {
+                    args.push(quote_ident!("arguments").as_arg());
+                }
+
                 Expr::Call(CallExpr {
                     span,
                     callee: fn_expr.make_member(quote_ident!("bind")).as_callee(),
-                    args: vec![ThisExpr { span: DUMMY_SP }.as_arg()],
+                    args,
                     type_args: Default::default(),
                 })
             }
-            _ => e,
+            _ => e.fold_children_with(self),
         }
     }
+
+    fn fold_constructor(&mut self, c: Constructor) -> Constructor {
+        let in_arrow = self.in_arrow;
+        self.in_arrow = false;
+        let res = c.fold_children_with(self);
+        self.in_arrow = in_arrow;
+        res
+    }
+
+    fn fold_function(&mut self, f: Function) -> Function {
+        let in_arrow = self.in_arrow;
+        self.in_arrow = false;
+        let res = f.fold_children_with(self);
+        self.in_arrow = in_arrow;
+        res
+    }
+}
+
+struct ArgumentsReplacer {
+    arguments: Ident,
+    found: bool,
+}
+
+impl Fold for ArgumentsReplacer {
+    noop_fold_type!();
+
+    fn fold_expr(&mut self, e: Expr) -> Expr {
+        match e {
+            Expr::Ident(id) if id.sym == js_word!("arguments") => {
```

Comment:
I think ideally we'd also check whether the syntax context is "global" here, but we don't have a reference to the global mark in this transform and adding one would be a breaking API change. This is only a problem in non-strict mode, because declarations with the name `arguments` are prohibited in strict mode. However, I couldn't figure out how to get SWC to parse in anything but strict mode (or at least, the error for `'eval' and 'arguments' cannot be used as a binding identifier in strict mode` always occurred). Is this something we need to worry about?

Reply:
A breaking change is ok I think. (Transforms need some breaking changes anyway).


It's strict mode because it's module code. 
https://swc.rs/docs/usage-core#transform 
This document is outdated, but it has `isModule: false,`.